aws/rust-runtime/aws-http/src/content_encoding.rs

```diff
@@ -0,0 +1,414 @@
+pub mod header_value {
+    /// Header value denoting "aws-chunked" encoding
+    pub const AWS_CHUNKED: &str = "aws-chunked";
+}
+
+/// Options used when constructing an [`AwsChunkedBody`][AwsChunkedBody].
+#[derive(Debug, Default)]
+#[non_exhaustive]
+pub struct AwsChunkedBodyOptions {
+    /// The total size of the stream. For unsigned encoding this implies that
+    /// there will only be a single chunk containing the underlying payload,
+    /// unless ChunkLength is also specified.
+    pub stream_length: Option<u64>,
+    /// The maximum size of each chunk to be sent.
+    ///
+    /// If ChunkLength and stream_length are both specified, the stream will be
+    /// broken up into chunk_length chunks. The encoded length of the aws-chunked
+    /// encoding can still be determined as long as all trailers, if any, have a
+    /// fixed length.
+    pub chunk_length: Option<u64>,
+    /// The length of each trailer sent within an `AwsChunkedBody`. Necessary in
+    /// order to correctly calculate the total size of the body accurately.
+    pub trailer_lens: Vec<u64>,
+}
+
+impl AwsChunkedBodyOptions {
+    /// Create a new [`AwsChunkedBodyOptions`][AwsChunkedBodyOptions]
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    /// Set stream length
+    pub fn with_stream_length(mut self, stream_length: u64) -> Self {
+        self.stream_length = Some(stream_length);
+        self
+    }
+
+    /// Set chunk length
+    pub fn with_chunk_length(mut self, chunk_length: u64) -> Self {
+        self.chunk_length = Some(chunk_length);
+        self
+    }
+
+    /// Set a trailer len
+    pub fn with_trailer_len(mut self, trailer_len: u64) -> Self {
+        self.trailer_lens.push(trailer_len);
+        self
+    }
+}
+
+#[derive(Debug, PartialEq, Eq)]
+enum AwsChunkedBodyState {
+    WritingChunkSize,
+    WritingChunk,
+    WritingTrailers,
+    Closed,
+}
+
+/// A request body compatible with `Content-Encoding: aws-chunked`
+///
+/// Chunked-Body grammar is defined in [ABNF] as:
+///
+/// ```txt
+/// Chunked-Body    = *chunk
+///                   last-chunk
+///                   chunked-trailer
+///                   CRLF
+///
+/// chunk           = chunk-size CRLF chunk-data CRLF
+/// chunk-size      = 1*HEXDIG
+/// last-chunk      = 1*("0") CRLF
+/// chunked-trailer = *( entity-header CRLF )
+/// entity-header   = field-name ":" OWS field-value OWS
+/// ```
+/// For more info on what the abbreviations mean, see https://datatracker.ietf.org/doc/html/rfc7230#section-1.2
+///
+/// [ABNF]:https://en.wikipedia.org/wiki/Augmented_Backus%E2%80%93Naur_form
+#[derive(Debug)]
+#[pin_project]
+pub struct AwsChunkedBody<InnerBody> {
+    #[pin]
+    inner: InnerBody,
+    #[pin]
+    state: AwsChunkedBodyState,
+    options: AwsChunkedBodyOptions,
+}
+
+// Currently, we only use this in terms of a streaming request body with checksum trailers
+type Inner = ChecksumBody<SdkBody>;
+
+impl AwsChunkedBody<Inner> {
```

Comment:
this is a little odd as an impl since we are hard coding this generic. Probably AwsChunkedBody should work even when not wrapping a checksum body?